beacon_node/beacon_chain/src/attestation_rewards.rs

```diff
@@ -0,0 +1,183 @@
+        mut validators: Vec<usize>,
+        log: Logger,
+    ) -> Result<StandardAttestationRewards, BeaconChainError> {
+        debug!(log, "computing attestation rewards"; "epoch" => epoch, "validator_count" => validators.len());
+
+        // Get state
+        let spec = &self.spec;
+
+        let state_slot = (epoch + 1).end_slot(T::EthSpec::slots_per_epoch());
+
+        let state_root = self
+            .state_root_at_slot(state_slot)?
+            .ok_or(BeaconChainError::UnableToFindTargetRoot(state_slot))?;
+
+        let state = self
+            .get_state(&state_root, Some(state_slot))
+            .and_then(|maybe_state| {
+                maybe_state.ok_or(BeaconChainError::MissingBeaconState(state_root))
+            })?;
+
+        // Calculate ideal_rewards
+        let participation_cache = ParticipationCache::new(&state, spec)?;
+
+        let previous_epoch = state.previous_epoch();
+
+        let mut ideal_rewards_hashmap = HashMap::new();
+
+        let penalty = 0;
+
+        for flag_index in PARTICIPATION_FLAG_WEIGHTS {
+            let weight = get_flag_weight(flag_index as usize)
+                .map_err(|_| BeaconChainError::AttestationRewardsSyncError)?;
+
+            let unslashed_participating_indices = participation_cache
+                .get_unslashed_participating_indices(flag_index as usize, previous_epoch)?;
+
+            let unslashed_participating_balance =
+                unslashed_participating_indices.total_balance().unwrap();
+
+            let unslashed_participating_increments =
+                unslashed_participating_balance.safe_div(spec.effective_balance_increment)?;
+
+            let total_active_balance = participation_cache.current_epoch_total_active_balance();
+
+            let active_increments =
+                total_active_balance.safe_div(spec.effective_balance_increment)?;
+
+            let base_reward_per_increment =
+                BaseRewardPerIncrement::new(total_active_balance, spec)?;
+
+            for effective_balance_eth in 0..=32 {
+                let base_reward =
+                    effective_balance_eth.safe_mul(base_reward_per_increment.as_u64())?;
+
+                let penalty =
+                    !0 * base_reward.safe_mul(weight)?.safe_div(WEIGHT_DENOMINATOR)? as i64;
+
+                let reward_numerator = base_reward
+                    .safe_mul(weight)?
+                    .safe_mul(unslashed_participating_increments)?;
+
+                let ideal_reward = reward_numerator
+                    .safe_div(active_increments)?
+                    .safe_div(WEIGHT_DENOMINATOR)?;
+                if !state.is_in_inactivity_leak(previous_epoch, spec) {
+                    ideal_rewards_hashmap
+                        .insert((flag_index, effective_balance_eth, penalty), ideal_reward);
+                } else {
+                    ideal_rewards_hashmap.insert((flag_index, effective_balance_eth, penalty), 0);
+                }
```

Comment:
Rather than putting the `penalty` in the key, which requires us to guess it correctly to look it up, lets put it as a value, like this:

```suggestion
                    ideal_rewards_hashmap
                        .insert((flag_index, effective_balance_eth), (ideal_reward, penalty));
                } else {
                    ideal_rewards_hashmap.insert((flag_index, effective_balance_eth), (0, penalty));
                }
```

This will require adjustments to the `fold` code and the total reward code.